gputop/remote/gputop-nodejs-csv.js

```diff
@@ -20,21 +24,124 @@ GPUTopNodeJSUI.prototype.weblog = function(message)
+    });
+
+    gputop.open_oa_metric_set({guid: features.supported_oa_query_guids[0]});
+    supported_oa_query_guids = features.supported_oa_query_guids;
+}
+
+GPUTopNodeJSUI.prototype.queue_redraw = function()
+{
+    if (supported_oa_query_guids !== undefined)
+        var metric = gputop.get_map_metric(supported_oa_query_guids[0]);
+}
+
+GPUTopNodeJSUI.prototype.load_metrics_panel = function(open_query)
+{
+    open_query();
+}
+
+GPUTopNodeJSUI.prototype.update_slider_period = function(period)
+{
+}
+
+GPUTopNodeJSUI.prototype.render_bars = function()
+{
+}
+
+GPUTopNodeJSUI.prototype.log = function(level, message)
+{
+    console.log(level);
     console.log(message);
 }
 
 var gputop_ui = new GPUTopNodeJSUI();
 
-function gputop_ready(gputop)
+function gputop_ready(gputop_inst)
 {
+    gputop = gputop_inst;
     gputop.connect();
 }
 
-http.get("http://localhost:7890/gputop.js", function(response) {
+function gputop_is_demo () {
+    return false;
+}
+
+function on_gputop_web_ready(gputop_web) {
+    http.get("http://localhost:7890/gputop.js", function(response) {
+        var gputopjs = "";
+        response.setEncoding('utf8');
+        response.on('data', function(data) {
+            gputopjs += data;
+        });
+        response.on('end', function () {
+            eval(gputop_web);
+            eval(gputopjs);
+        });
+    });
+}
+
+http.get("http://localhost:7890/gputop-web.js", function(response) {
+    var gputop_web = "";
     response.setEncoding('utf8');
     response.on('data', function(data) {
-        eval(data);
+        gputop_web += data;
+    });
+    response.on('end', function () {
+        on_gputop_web_ready(gputop_web);
     });
 });
+
+function gputop_clean_exit()
+{
+    if (gputop !== undefined && supported_oa_query_guids !== undefined) {
+        var metric = gputop.get_map_metric(supported_oa_query_guids[0]);
+        var fs = require('fs');
+        var stream = fs.createWriteStream("my_file.csv");
+        stream.once('open', function(fd) {
+            var csv_file = "Metric,Counter,Start,End,Value,Maximum,Reason\n";
+
+            for (j = 0; j < metric.counters_.length; j++) {
+                for (i = 0; i < metric.counters_[j].updates.length; i++) {
+                    csv_file += "" + metric.name_ + "," +
+                        metric.counters_[j].symbol_name + "," +
+                        metric.counters_[j].updates[i][0] + "," +
+                        metric.counters_[j].updates[i][1] + "," +
+                        metric.counters_[j].updates[i][2] + "," +
+                        metric.counters_[j].updates[i][3] + "," +
+                        metric.counters_[j].updates[i][4] + "," + "\n";
+                }
+            }
+
+            stream.write(csv_file);
+            stream.end();
+            pkill('gputop', function(error, valid_pid) {
+                if (error)
+                    console.log(error);
+                console.log(valid_pid);
+            });
```

Comment:
maybe as a later iteration, I think we should aim to add a protocol buffer request for closing the server and also when running a node.js client we could have the server run in a mode where it knows to quit if the connection is lost.

using pkill won't work in cases where the server is remote which I'd like to support.